core/src/main/java/com/linecorp/armeria/common/RequestContextAwareCompletableFuture.java

```diff
@@ -243,4 +243,92 @@
     public CompletableFuture<T> exceptionally(Function<Throwable, ? extends T> fn) {
         return ctx.makeContextAware(super.exceptionally(ctx.makeContextAware(fn)));
     }
+
+    // support JDK9 functions
+
+    public <U> CompletableFuture<U> newIncompleteFuture() {
+        return new RequestContextAwareCompletableFuture<>(ctx);
+    }
+
+    public CompletionStage<T> minimalCompletionStage() {
+        return this.uniAsMinimalStage();
+    }
+
+    public CompletableFuture<T> completeAsync(Supplier<? extends T> supplier) {
+        return ctx.makeContextAware(supplyAsync(this.makeContextAware(supplier)));
+    }
+
+    public CompletableFuture<T> completeAsync(Supplier<? extends T> supplier,
+                                              Executor executor) {
+        return ctx.makeContextAware(supplyAsync(this.makeContextAware(supplier), executor));
+    }
+
+    private Supplier<T> makeContextAware(Supplier<? extends T> action) {
+        return () -> {
+            try (SafeCloseable ignored = ctx.pushIfAbsent()) {
+                return action.get();
+            }
+        };
+    }
+
+    private MinimalStage<T> uniAsMinimalStage() {
+        return new MinimalStage<>();
+    }
+
+    static final class MinimalStage<T> extends CompletableFuture<T> {
```

Comment:
This isn't context aware since it doesn't extend `RequestContextAwareCompletableFuture` or return one. How about

```java
final class MinimalStage<T> extends RequestContextAwareCompletableFuture<T> {
  MinimalStage() {
    super(ctx);
    RequestContextAwareCompletableFuture.this.whenComplete((value, t) -> {
      if (t != null) {
        super.completeExceptionally(t);
      } else {
        super.complete(value);
   }
}
```

Reply:
Also given how complicated this API is, I think we can just `throw new UnsupportedOperationException` for now in `minimalCompletionStage()` - we could revisit after copying in the `CompletableFuture` tck so it's easier to verify the behavior works as the JDK expects, and in the meantime I think it's a very uncommon API.

Reply:
I guess it's not allowed to fork the GPL'd TCK into our code base. I guess we can implement the `MinimalStage` by delegating all supported operations to `RequestContextAwareCompletableFuture` and wrapping the returned `CompletableFuture` with `MinimalStage`:

```java
final class MinimalStage<T> implements CompletionStage<T> {
    private final RequestContextAwareCompletableFuture<T> delegate;

    public CompletionStage<Void> thenRun(...) {
        return new MinimalStage<>(delegate.thenRun(...));
    }
    public boolean complete(...) {
        throw new UnsupportedOperationException();
    }
    public CompletableFuture toCompletableFuture() {
        return delegate.thenApply(Function.identity());
    }
}
```